### PR TITLE
Checkout feature

### PR DESCRIPTION
**Pull request by KevinLay689**
### Summary of changes
- Adding support here for native checkout
- There are a few core things that we need for checkout to start
  1. Correlation ID - this tells us how to associate a checkout session throughout the call
  2. ReturnUrl - this tells the webview where to return to when auth is complete, eventually we will remove the need for this and use a universal return url
  3. BrainTree Account that is linked with a PayPal account so that a valid client ID is returned from the configuration objects payload.

After the `PaypalRequest` object is made, the `tokenize` call on the `PayPalNativeCheckoutClient` will launch either a BA session or single payment session.

Our core logic for starting the session is in this block:

```
private void sendPayPalRequest(
        final FragmentActivity activity,
        final PayPalRequest payPalRequest,
        final Configuration configuration,
        final PayPalFlowStartedCallback callback
    ) {
        internalPayPalClient.sendRequest(activity, payPalRequest, (payPalResponse, error) -> {
            if (payPalResponse != null) {
                String analyticsPrefix = getAnalyticsEventPrefix(payPalRequest);
                braintreeClient.sendAnalyticsEvent(String.format("%s.app-switch.started", analyticsPrefix));

                Environment environment;
                if ("sandbox".equals(configuration.getEnvironment())) {
                    environment = Environment.SANDBOX;
                } else if ("production".equals(configuration.getEnvironment())) {
                    environment = Environment.LIVE;
                } else {
                    callback.onResult(new IllegalArgumentException("Invalid PayPal Environment"));
                    return;
                }

                // Start PayPalCheckout flow
                PayPalCheckout.setConfig(
                    new CheckoutConfig(
                        activity.getApplication(),
                        configuration.getPayPalClientId(),
                        environment,
                        payPalRequest.getNativeConfig().getReturnUrl()
                    )
                );

                registerCallbacks(configuration, payPalRequest, payPalResponse);

                PayPalCheckout.startCheckout(createOrderActions -> {
                    if (payPalRequest instanceof PayPalCheckoutRequest) {
                        createOrderActions.set(payPalResponse.getPairingId());
                    } else if (payPalRequest instanceof PayPalVaultRequest) {
                        createOrderActions.setBillingAgreementId(payPalResponse.getPairingId());
                    } else {
                        callback.onResult(new Exception("Unsupported action type"));
                    }
                });
                callback.onResult(null);
            } else {
                callback.onResult(error);
            }
        });
    }
```

### Video

https://user-images.githubusercontent.com/15115314/173677141-4627492c-b2a2-49a8-9fa3-cf503d38e5e5.mov

 ### Checklist

 - [ ] Added a changelog entry

-

### CODE REVIEW

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (KevinLay689) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:266]: This is where we begin the checkout flow

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (KevinLay689) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:298]: @tdchow do you think we should allow the user to pass their own callbacks here? Originally I was thinking we should but we need to control the `internalPayPalClient.tokenize` call here

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (KevinLay689) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:322]: So this is a really basic version of the one that was provided here: https://github.com/braintree/braintree_android/blob/00b21449e3fee7595c29ce2c5e31001db7439eb1/PayPal/src/main/java/com/braintreepayments/api/PayPalClient.java#L314

I could potentially need to add more but not sure where to get the rest of these params cause 
```
        final Configuration configuration,
        final PayPalRequest payPalRequest,
        final PayPalResponse payPalResponse
```
does not contain the same values as `browserSwitchResult.getRequestMetadata();`

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (KevinLay689) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:326]: can I use this client ID? or should this be the correlationID here? @tdchow

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutConfig.java
Comment (KevinLay689) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutConfig.java:6]: I just added this to be safe in case we wanted to support adding in the 3p callbacks here later. I believe both of these things could be moved or fetched elsewhere as well

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (tdchow) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:298]: I believe this would be up to the BT team on what their external interface would look like. Would we want to keep this new native PayPal interface matching the others?

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (tdchow) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:326]: I think it would be what you have here. I believe the merchant's clientID would be connected to the BT account somehow, and the string value of the clientID would be returned in the BT configuration call. I may be completely off here, but please someone from BT chime in if you can!

**Review by tdchow — APPROVED** [review]
Overall things look good on my end, just had a few questions. I also think we need to wait for someone from the BT team to review before merging!

File: Demo/src/main/java/com/braintreepayments/demo/PayPalRequestFactory.java
Comment (sshropshire) [inline Demo/src/main/java/com/braintreepayments/demo/PayPalRequestFactory.java:48]: Can we encapsulate this logic within `PayPalNativeRequest` and add a `returnUrl` property to simplify the interface?

File: GooglePay/src/main/java/com/braintreepayments/api/GooglePayRequest.java
Comment (sshropshire) [inline GooglePay/src/main/java/com/braintreepayments/api/GooglePayRequest.java:199]: We should create a separate PR for these Google Pay changes.

File: GooglePay/src/test/java/com/braintreepayments/api/GooglePayRequestUnitTest.java
Comment (sshropshire) [inline GooglePay/src/test/java/com/braintreepayments/api/GooglePayRequestUnitTest.java:118]: +1 for separate PR.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalAccountNonce.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalAccountNonce.java:16]: +1 naming collison.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalBrowserSwitchException.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalBrowserSwitchException.java:6]: +1 naming collison.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalCheckoutRequest.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalCheckoutRequest.java:21]: +1 naming collison.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalCreditFinancing.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalCreditFinancing.java:15]: +1 naming collison.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalCreditFinancingAmount.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalCreditFinancingAmount.java:12]: +1 naming collison.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalFlowStartedCallback.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalFlowStartedCallback.java:12]: +1 naming collison.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalInternalClient.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalInternalClient.java:12]: The package private classes may collide. I'm not 100% sure, but we should rename to be safe.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalInternalClientCallback.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalInternalClientCallback.java:5]: +1 potential naming collision.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalLineItem.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalLineItem.java:16]: +1 naming collision.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalListener.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalListener.java:8]: +1 naming collision.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (sshropshire) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:321]: A lot of the values come from the callback URL in the OG PayPal module. Are we missing some values?

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (sshropshire) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:325]: Yeah whichever values we're using in the PayPal web module should be fine.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (sshropshire) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:327]: I'm not sure either we should follow up. We've only had one flow up until this point.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:371]: Do we still need this code?

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutConfig.java
Comment (sshropshire) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutConfig.java:6]: Can we remove this for now and add later?

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalPaymentIntent.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalPaymentIntent.java:13]: +1 naming collision.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalPaymentResource.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalPaymentResource.java:9]: +1 potential naming collision.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalRequest.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalRequest.java:21]: +1 naming collision.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalResponse.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalResponse.java:3]: +1 potential naming collision.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalVaultRequest.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalVaultRequest.java:13]: +1 naming collision.

File: GooglePay/src/main/java/com/braintreepayments/api/GooglePayRequest.java
Comment (KevinLay689) [reply GooglePay/src/main/java/com/braintreepayments/api/GooglePayRequest.java:199]: whoops looks like auto-formatting got it

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/BTPayPalNativeOrderCreationClient.java
Comment (KevinLay689) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/BTPayPalNativeOrderCreationClient.java]: deleted class

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalAccount.java
Comment (KevinLay689) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalAccount.java]: Yeah good callout, ill prefix it all with `PayPalNativeCheckout`

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (KevinLay689) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java]: I'm actually not sure, there seems to be a few things inside `getRequestMetadata();` that aren't inside the other requests. I don't think its necessary to perform the checkout but won't know unless the nonce value succeeds

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutConfig.java
Comment (KevinLay689) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutConfig.java]: yeah we can remove this, we can use the `PayPalNativeCheckoutRequest` to hold these for now

File: Demo/src/main/java/com/braintreepayments/demo/PayPalRequestFactory.java
Comment (KevinLay689) [reply Demo/src/main/java/com/braintreepayments/demo/PayPalRequestFactory.java]: yup, done.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalAccountNonce.java
Comment (KevinLay689) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalAccountNonce.java]: Resolved all the naming collisions

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (KevinLay689) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:291]: Ok, we can defer adding the callbacks for now then

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (jaxdesmarais) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:173]: I'm assuming these deprecated methods are an android thing or can these all be removed?

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (jaxdesmarais) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:109]: If these are new exceptions can we change the docs link to the correct link here or do we need to keep the legacy URLs in this new flow?

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (jaxdesmarais) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:320]: Yeah I assume we'd want to set something like `paypal-native-checkout`

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (sshropshire) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:173]: This is a good callout. We should simplify the interface where we can. It is `next-major-version` material in the OG PayPal module but since this is new we should remove deprecated API.

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (KevinLay689) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java]: No we don't, removed all the browser switch code

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (KevinLay689) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:173]: Removed the deprecated methods

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (KevinLay689) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:320]: So when I tried changing this I received an error when making the request.. do we know if it this is like a whitelist somewhere?

File: Demo/src/main/java/com/braintreepayments/demo/PayPalNativeCheckoutFragment.java
Comment (sshropshire) [inline Demo/src/main/java/com/braintreepayments/demo/PayPalNativeCheckoutFragment.java:74]: +1

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (sshropshire) [inline PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:186]:
```suggestion
        return request instanceof PayPalNativeCheckoutVaultRequest ? "paypal-native.billing-agreement" : "paypal-native.single-payment";
```

File: build.gradle
Comment (sshropshire) [reply build.gradle:15]: No worries. On an M1 too I know the pain lol.

File: Demo/src/main/java/com/braintreepayments/demo/PayPalNativeCheckoutFragment.java
Comment (KevinLay689) [reply Demo/src/main/java/com/braintreepayments/demo/PayPalNativeCheckoutFragment.java:67]: Done

File: Demo/src/main/java/com/braintreepayments/demo/PayPalNativeCheckoutFragment.java
Comment (KevinLay689) [reply Demo/src/main/java/com/braintreepayments/demo/PayPalNativeCheckoutFragment.java:74]: Done

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (KevinLay689) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java]: Done, switched events to use `paypal-native`

File: PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java
Comment (jaxdesmarais) [reply PayPalNativeCheckout/src/main/java/com/braintreepayments/api/PayPalNativeCheckoutClient.java:173]: Agree - maybe paypal or paypal-app, would be nice to get paypal-native-checkout added but that isn't a blocker.

**Review by jaxdesmarais — APPROVED** [review]
Thanks for getting this feature added and addressing comments! 🥳

**Comment by jcnoriega** [comment]
LGTM!

**Pull request merged**
